src/NavigationHelper/src/NavigationHelper.cpp

```diff
@@ -0,0 +1,171 @@
+/**
+ * @file NavigationHelper.cpp
+ * @authors Simone Micheletti <simone.micheletti@iit.it>
+ * @copyright 2023 iCub Facility - Istituto Italiano di Tecnologia
+ *            Released under the terms of the LGPLv2.1 or later, see LGPL.TXT
+ * @date 2023
+ */
+
+#include <iostream>
+#include <thread>
+#include <WalkingControllers/NavigationHelper/NavigationHelper.h>
+
+#include <WalkingControllers/YarpUtilities/Helper.h>
+
+
+using namespace WalkingControllers;
+
+NavigationHelper::NavigationHelper()
+{
+}
+
+NavigationHelper::~NavigationHelper()
+{
+}
+
+bool NavigationHelper::closeThreads()
+{
+    //Close parallel threads
+    if (m_runThreads)
+    {
+        m_runThreads = false;
+
+        yInfo() << "Closing m_navigationTriggerThread";
+        if(m_navigationTriggerThread.joinable())
+        {
+            m_navigationTriggerThread.join();
+            m_navigationTriggerThread = std::thread();
+        }
+    }
+    return true;
+}
+
+bool NavigationHelper::closeHelper()
+{
+    try
+    {
+        //close threads first
+        if (m_runThreads)
+        {
+            closeThreads();
+        }
+        //close ports
+        if(!m_replanningTriggerPort.isClosed())
+            m_replanningTriggerPort.close();
+    }
+    catch(const std::exception& e)
+    {
+        std::cerr << e.what() << std::endl;
+        return false;
+    }
+    return true;
+}
+
+bool NavigationHelper::init(const yarp::os::Searchable& config, std::deque<bool> &leftInContact, std::deque<bool> &rightInContact )
+{
+    m_leftInContact = &leftInContact;
+    m_rightInContact = &rightInContact;
```

Comment:
This is not thread-safe. You are taking the pointer of some objects, and then you access the objects in a different thread. You might have concurrent read and write access to these variables and this is an undefined behavior.

One suggestion could be having internal copies of these variables that are updated from outside (i.e. from the ``Module``), and using mutexes to access them.

Reply:
Ok I have changed that, but I don't know where in the `Module.cpp` I can update them, calling that access method.

Reply:
> Maybe in `updateTrajectories`?

Sounds good!

Reply:
Ok, I have also added a `m_navHelperUsed` flag to check before calling the update method.
This flag will be set to true if `navigationOptions` will not be null